// node-package/dist/success.view.js
import { html, svg } from '@nutmeg/seed';
import { repeat } from 'lit-html/directives/repeat';
export class SuccessView {
    constructor(component, pkg, selectedInstallCommand) {
        this.component = component;
        this.pkg = pkg;
        this.selectedInstallCommand = selectedInstallCommand;
    }
    get name() {
        return this.pkg.name;
    }
    get content() {
        return html `
      <div id="description" class="row">
        ${this.pkg.description}
      </div>
      ${this.keywords}
      ${this.install}
      ${this.footer}
      <div id="toast-wrapper">
        <span id="toast">Copied to clipboard</span>
      </div>
    `;
    }
    selectInstallCommand(event, command) {
        event.preventDefault();
        this.component.selectedTab = command.id;
        this.component.render();
    }
    copyInstallCommand(event) {
        event.preventDefault();
        this.component.$('.command:not(.hidden)').select();
        document.execCommand('copy');
        this.component.$('#toast').classList.add('copied');
        setTimeout(() => {
            this.component.$('#toast').classList.remove('copied');
        }, 2750);
    }
    keyword(keyword) {
        return html `
        <a href="https://www.npmjs.com/browse/keyword/${keyword}" target="_blank" rel="noopener" class="keyword">#${keyword}</a>
    `;
    }
    get keywords() {
        return html `
      <div id="keywords" class="row ellipsis">
        ${repeat(this.pkg.keywords, keyword => keyword, (keyword, _index) => this.keyword(keyword))}
      </div>
    `;
    }
    get types() {
        return html `
      <span class="item" title="${this.pkg.types}">Includes types</span>
    `;
    }
    installTab(command) {
        const classes = `item tab ${this.selectedInstallCommand === command.id ? 'selected' : ''}`;
        return html `
      <a class="${classes}" href="#" @click=${(event) => this.selectInstallCommand(event, command)}>
        <label for="${command.id}">
          ${command.id}
        </label>
      </a>
    `;
    }
    installCommand(command) {
        return html `
      <input id="${command.id}" class="command fixed-width ellipsis item ${this.selectedInstallCommand !== command.id ? 'hidden' : ''}" readonly value="${command.command}">
    `;
    }
    get install() {
        return html `
      <div id="install" class="row">
        <div id="tabs">
          ${repeat(this.pkg.installCommands(this.component.global), command => command.id, (command, _index) => this.installTab(command))}
        </div>
        <div id="commands" class="row-horizontal">
          ${repeat(this.pkg.installCommands(this.component.global), command => command.id, (command, _index) => this.installCommand(command))}
          <div class="item">
            <a id="copy" href="#" title="Copy command" @click=${(event) => this.copyInstallCommand(event)}>${this.copy}</a>
          </div>
        </div>
      </div>
    `;
    }
    get footer() {
        return html `
      <div id="footer" class="row row-horizontal">
        <span class="item">
          v${this.pkg.version}
        </span>
        ${this.pkg.types && this.types}
        <span class="item">
          ${this.pkg.license}
        </span>
      </div>
    `;
    }
    get copy() {
        return svg `
      <svg fill="#000000" height="24" viewBox="0 0 24 24" width="24" xmlns="http://www.w3.org/2000/svg">
        <path d="M0 0h24v24H0z" fill="none"/>
        <path d="M16 1H4c-1.1 0-2 .9-2 2v14h2V3h12V1zm3 4H8c-1.1 0-2 .9-2 2v14c0 1.1.9 2 2 2h11c1.1 0 2-.9 2-2V7c0-1.1-.9-2-2-2zm0 16H8V7h11v14z"/>
      </svg>
    `;
    }
}
